creusot/src/backend/ty.rs

```diff
@@ -410,14 +410,31 @@ fn field_ty<'tcx>(
     ctx: &mut Why3Generator<'tcx>,
     names: &mut CloneMap<'tcx>,
     param_env: ParamEnv<'tcx>,
+    adt_did: DefId,
     field: &FieldDef,
     substs: SubstsRef<'tcx>,
 ) -> MlT {
     let ty = field.ty(ctx.tcx, substs);
     let ty = ctx.try_normalize_erasing_regions(param_env, ty).unwrap_or(ty);
+
+    if !validate_field_ty(ctx.tcx, adt_did, ty) {
+        ctx.crash_and_error(ctx.def_span(field.did), "Illegal use of the Ghost type")
+    }
+
     translate_ty_inner(TyTranslation::Declaration, ctx, names, ctx.def_span(field.did), ty)
 }
 
+fn validate_field_ty<'tcx>(tcx: TyCtxt<'tcx>, adt_did: DefId, ty: Ty<'tcx>) -> bool {
+    !ty.walk().filter_map(ty::GenericArg::as_type).any(|ty| {
+        util::is_ghost_ty(tcx, ty)
+            && ty.walk().filter_map(ty::GenericArg::as_type).any(|ty| match ty.kind() {
+                TyKind::Adt(adt_def, _) => adt_def.did() == adt_did,
+                // TyKind::Param(_) => true,
```

Comment:
we should also disallow `Ghost<T>`, however, this breaks e.g. `MapInv`which has a field `produced: Ghost<Seq<B>>`.

Reply:
This seems stricter than necessary, ideally, it would work something like this:
```rust
struct HasGhost<T, U>(Ghost<T>, U); // No Error

struct RecurFine(HasGhost<u32, RecurFine>); // No Error

struct RecurBad(HasGhost<RecurBad, u32>); // Error
```

This would probably require tracking for each ADT which type parameters flow through `Ghost`.

Reply:
Agreed, and we discussed doing just that in person but the idea is to start with the fastest, most restrictive and sound version first.